Toy v1 OpenQL v1/v3 reader

rturrado opened the pull request:
OpenQL update to work with the latest libqasm.

The latest libqasm:
- adds ANTLR for parsing cQASM 3.0 files, and
- builds only via Conan.

This commit brings the following changes to OpenQL:
- Can only be built via Conan.
- Works with the latest libqasm.
- Changes the Reader's `read` function to enroute parsing of cQASM 3.0 files to the libqasm's v3 parser (ANTLR).
- Fixes all the GitHub Actions builds.

rturrado committed:
cpp_standalone_example is now a build option.
It may be removed in the future, as simply by building a test executable we are already proving what cpp_standalone_example aims at.

rturrado committed:
Fixed test.yml and CMakeLists.txt.
test.yml:
- Added 'working-directory' to 'Run tests and standalone program' step.
- Added 'conan' to the list of python dependencies.
CMakeLists.txt:
- Changed 'cmake -E copy -t' to 'cmake -E copy' since '-t' is only available from CMake 3.26.

rturrado committed:
Added 'conan profile detect' to the Build step of the Python job.

rturrado committed:
CMakeLists.txt: added policy checks.
openql.i: defined docstring in multiple lines.
setup.py: added OPENQL_BUILD_TESTS and OPENQL_BUILD_CPP_STANDALONE_EXAMPLE options.
test.yml:
- Updated actions' versions.
- Added ' --use-pep517' to 'python pip' install line.
- Removed comment regarding Conda.

rturrado committed:
test.yml: temporarily commented Cpp job, as well as macOS,Windows, and python 3.7 to 3.10 Python job builds.
setup.py: temporarily commented code that does CMake configure, and build.

rturrado committed:
setup.py: updated to do the installation step via CMake.

rturrado committed:
Merge branch 'develop' into v3

# Conflicts:
#	.github/workflows/test.yml
#	CMakeLists.txt
#	conanfile.py
#	python/openql.i
#	setup.py

rturrado committed:
Updated dependencies.

rturrado committed:
Updated setup.py and test.yml.

rturrado committed:
Trying to fix the Conan builds.
- conanfile.py: updated Conan options, and matched them to CMake options.
- setup.py: added Conan requirement.
- test.yml: fixed Conan line for cpp-arm64 job.

rturrado committed:
Trying to fix the Conan builds.
- CMakeLists.txt: updated to the latest libqasm commit.
- conanfile.py: updated Conan package name from 'ql' to 'openql'.
- test.yml: added Conan to Python dependencies for 'python' job.

rturrado committed:
Trying to fix the Conan builds.
- CMakeLists.txt: reverted change that was updating libqasm to the latest commit. That should be done in OpenQL/v3_1.
- python/CMakeLists.txt: removed add_backward line.
- conanfile.py: added exports_sources, package, and package_info.
- LICENSE: Renamed to LICENSE.md.
- setup.py: added build_tests and disable_unitary local variables.

rturrado committed:
Trying to fix the Conan builds:
- conanfile.py:
  - Moved layout after requirements.
  - Most importantly, changed the layout code to mimic that of libqasm, instead of using the default layout. Among other things, this allows us to set the build folder to 'build/<build_type>', which is not happening for Windows.
  - Also, fixed the name of WITH_UNITARY_DECOMPOSITION.
- setup.py: Darwin builds need '-D_LIBCPP_DISABLE_AVAILABILITY'.

rturrado committed:
Trying to fix ARM compilations.

rturrado committed:
Updated namespace definitions in read.cc and cqasm_reader.h.

rturrado committed:
Updated highs Conan package to version 1.5.3.

rturrado committed:
Trying to fix the Conan builds.
- test.yml: updated 'on-push-branches'.
- CMakeLists.txt: updated to the latest libqasm commit.
- include/read.h: added read_v3 function.
- src/read.cc: added read_v3 function, and changed 'read' function to enroute whether to read_v1 or read_v3 depending on the parsed version.
- test/read.cc: updated check for parsing of v3 files.

rturrado committed:
Updated libqasm version.
OpenQL wasn't able to find the Python executable.

rturrado committed:
Changed 'read' function so that it distinguishes between versions up to 1.2, version 3.0, and others.
Added fmt library requirement, then used it in 'read' function fo format the error message.

rturrado committed:
Updated to latest libqasm/develop commit.

rturrado committed:
Updated 'on-push' to point to 'develop'.

pablolh reviewed (commented):
Nice!
I think it might need some modifications to the docs/ folder and the README.md because they probably still refer to CMake and submodules?

rturrado commented:
> Nice! I think it might need some modifications to the docs/ folder and the README.md because they probably still refer to CMake and submodules?

Hey Pablo! Yes, very good point! I definitely have to revise the README and the documentation in general.

rturrado committed:
Updated docs. [skip ci]

rturrado committed:
Merge remote-tracking branch 'origin/toy-v1-OpenQL_v1_v3_reader' into toy-v1-OpenQL_v1_v3_reader

rturrado committed:
Updated docs. [skip ci]
- Updated `sphinx` version so that local build works.
- Fixed syntax for nested lists.

rturrado committed:
Just a silly commit to force GitHub Actions to run.
- We have to keep an eye on test_condex.py because it doesn't finish quite often for Python/MacOS builds.
- This commit introduces just minor aesthetic changes in test_condex.py.
